homeassistant/components/opensprinkler/__init__.py

```diff
@@ -0,0 +1,177 @@
+"""Opensprinkler integration."""
+import asyncio
+from datetime import timedelta
+import logging
+
+from pyopensprinkler import OpenSprinkler
+import voluptuous as vol
+
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PASSWORD, CONF_PORT
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.util import Throttle
+
+from .const import (
+    CONF_RUN_SECONDS,
+    DATA_DEVICES,
+    DEFAULT_NAME,
+    DEFAULT_PORT,
+    DEFAULT_RUN_SECONDS,
+    DOMAIN,
+    SCAN_INTERVAL,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_DEVICES = "devices"
+PLATFORMS = ["sensor"]
+
+DEVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_RUN_SECONDS, default=DEFAULT_RUN_SECONDS): cv.positive_int,
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {vol.Required(CONF_DEVICES): vol.All(cv.ensure_list, [DEVICE_SCHEMA])}
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the opensprinkler component from YAML."""
+
+    conf = config.get(DOMAIN)
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN].setdefault(DATA_DEVICES, {})
+
+    if not conf:
+        return True
+
+    for device_config in conf[CONF_DEVICES]:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=device_config
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Opensprinkler from a config entry."""
+    password = entry.data.get(CONF_PASSWORD)
+    host = f"{entry.data.get(CONF_HOST)}:{entry.data.get(CONF_PORT, DEFAULT_PORT)}"
+    hass.data[DOMAIN][entry.entry_id] = await hass.async_add_executor_job(
+        OpenSprinkler, host, password
+    )
```

Comment:
Will an exception be raised if the connection cannot be established or can you check if auth is still valid? If connection cannot be established right now, raise ConfigEntryNotReady and Home Assistant will retry in a bit.